### PR TITLE
feat(mgmt): sort method parameters

### LINKED ISSUE #1843: Sort parameters ordering
[1771](https://github.com/Azure/autorest.csharp/issues/1771)

We need to sort the 2, 3 in above issue

### PR DESCRIPTION
**Pull request by archerzz**
# Description
For mgmt method parameters, we've decided the new order is:
1. `waitForCompletion` (LRO only)
2. required parameters
3. body parameters
4. optiona parameters
5. `cancellationToken`

Item 1 and 5 are not swagger definitions, and they will be added by generator automatically. So we only need to implement the sorting for 2, 3, 4 above.
 This commit:
- update the logic of parameter sorting
- refactor to rename `OrderParameters` to `OrderParametersByRequired`, since it won't ensure body parameter position
- add test cases

resolve #1843

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs [hunk 1]
```diff
@@ -289,17 +293,38 @@ private Request BuildRequest(HttpRequest httpRequest, RequestMethodBuildContext
...
+            List<Parameter> requiredParameters = new();
+            List<Parameter> optionalParameters = new();
+            List<Parameter> bodyParameters = new();
             foreach (var (requestParameter, parameter) in allParameters)
             {
                 // Grouped and flattened parameters shouldn't be added to methods
                 if (IsMethodParameter(requestParameter))
                 {
-                    methodParameters.Add(parameter);
+                    // sort the parameters by the following sequence:
```
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:304]: This change will break parameter ordering in all existing HLC clients.

**Review by AlexanderSher — DISMISSED** [review]
This is a heavy breaking change. What issue this PR is addressing?
If this is a fix to https://github.com/Azure/autorest.csharp/issues/1843, then it should be scoped to mgmt plane. You can also try to use `RestClientBuilder.BuildRequestMethod` that does correct sorting for LLC.

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs (hunk 1 above)
Comment (archerzz) [reply src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:304]: Thanks. I thought the ordering is for HLC as well. I'll split the logic.

**Comment by archerzz** [comment]
@AlexanderSher I updated the codes to split the sorting logic for mgmt and data plane. The logic for mgmt is much simpler than LLC, so I still keep its own version of sorting.

**Comment by AlexanderSher** [comment]
I remember that we have discussed that order of host and path parameters should match their order in the URI template. Is it handled in some other place in Mgmt?

**Comment by archerzz** [comment]
> I remember that we have discussed that order of host and path parameters should match their order in the URI template. Is it handled in some other place in Mgmt?

@AlexanderSher Yes. Path parameters are sorted in https://github.com/Azure/autorest.csharp/blob/faf9247db104799f62d518f78942164102d83408/src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs#L770

As for `host`, do you mean `x-ms-parameterized-hos` in swagger? That seems not exist in mgmt swagger, so we don't handle it.

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs (hunk 1 above)
Comment (ArthurMa1978) [reply src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:332]: according to https://github.com/Azure/autorest.csharp/issues/1771 this should apply to all .net sdk

**Review by m-nash — APPROVED** [review]

**Pull request merged**